buildsys: variant and readme standardization

webern opened the pull request:
**Issue number:**

Related to #1260
Also will be used for #1419
Also see https://github.com/bottlerocket-os/bottlerocket/pull/2098#discussion_r859041469

**Description of changes:**

```
generate readme

We had readme generating code pasted into most of our build.rs file.
This has now been moved to a public function in a library so that the
code is not duplicated.
```

```
standardize variant concepts

We had several crates that interpreted the VARIANT env variable in
different ways in build.rs. This standardizes the treatment of the
VARIANT variable by centralizing it in a library.
```

**Ain't nobody got time for that**

- First look at `tools/buildsys/src/public/readme.rs`
- Then look at, e.g. `sources/shimpei/build.rs`: simple, we've just refactored away the duplicative readme generation code.
- Next look at `tools/buildsys/src/public/variant.rs`
- Then look at, e.g. `sources/api/pluto/build.rs`: we have factored out the emitting of `cfg` variables in the build.rs

That's basically it. Most of the crates are copy-paste nothing interesting. But these crates are the ones that actually have `cfg` usage:
- `sources/api/early-boot-config/build.rs`
- `sources/api/ecs-settings-applier/build.rs`
- `sources/api/pluto/build.rs`
- `sources/api/static-pods/build.rs`
- `sources/api/storewolf/build.rs`
- `sources/logdog/build.rs`
- `sources/models/build.rs`

**Testing done:**

- [x] I changed a `README.tpl` file and observed that `README.md` was changed.
- [x] I changed a rust doc comment and observed that `README.md` was changed.
- [x] I verified that SKIP_README skips readme generation.
- [x] run an aws-k8s sonobuoy 'quick'
- [x] run an ecs task
- [x] run a vmware sonobuoy quick

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

webern commented:
^ fix a mistake in pluto's build.rs and improve a doc comment

webern commented:
^ fix a couple of mistakes

webern commented:
^ rebase with fixed conflicts

webern commented:
^ improve some doc comments

webern commented:
:cricket: :cricket: :cricket: 

@zmrow @cbgbt can someone get the ball rolling on this PR? I'll try to do VMWare testing today.

webern commented:
^ rebase

zmrow commented on sources/api/ecs-settings-applier/build.rs:
```diff
@@ -1,26 +1,15 @@
-        println!("cargo:rustc-cfg=variant_type=\"{}\"", variant_type);
-    }
+    // This build has variant-specific logic.
+    let variant = Variant::from_env().unwrap();
```
This `unwrap()` made me a little sad initially, but on further thought I think it might be fine since it's during the build and should print something intelligible.

webern replied:
> This unwrap() made me a little sad initially, but on further thought I think it might be fine since it's during the build and should print something intelligible.

Yeah I think it's normal to panic in build.rs. I suppose technically we could `eprintln!` and then `std::process::exit(1)` but we weren't doing that before.

zmrow reviewed (approved):
lgtm

:nut_and_bolt:

bcressey commented on Dockerfile:
```diff
@@ -114,6 +114,7 @@ RUN --mount=source=.cargo,target=/home/builder/.cargo \
     --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/models/src/variant,from=variantcache,source=/variantcache \
     --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/logdog/conf/current,from=variantcache,source=/variantcache \
     --mount=source=sources,target=/home/builder/rpmbuild/BUILD/sources \
+    --mount=source=tools,target=/home/builder/rpmbuild/BUILD/tools \
```
This crosses a line that might exist only in my mind, but is still worth considering:
* `sources` for everything that might ship in an OS image
* `tools` for the disjoint set of everything needed to build OS images

This has some symmetry with the host vs. target distinction that cross-compilation introduces.

Things that could be independent crates (`parse-datetime`) that are useful to both are placed in sources until they graduate to their new home.

I'd prefer to follow that same pattern by adding a `generate-readme` crate to `sources` and then using that everywhere, including in `tools`.

(Eventually, in pursuit of faster build times, I could see pushing everything in `tools` into the SDK instead, at which point we'd need the shared dependencies to become external crates.)

webern replied:
`generate_readme` can be put into `sources` no problem.

The challenge with `Variant` is that we want to use it in both `tools` and `sources`.

Option 1:

If we put it into `sources` then we can reference it from `tools`. We have references already from `tools` -> `sources` already, so this is precedented https://github.com/bottlerocket-os/bottlerocket/blob/a7816b921cbb734a33e66d9e8fc95584fdb9c281/tools/pubsys/Cargo.toml#L46

Option 2:

We could start a new workspace at the top level, e.g. `shared`, for types that are going to be used by both `tools` and `sources` which would make crates used by both tools and sources visually apparent.

webern replied:
P.S. in re-reading I see you already hinted at Option 2. So probably what we do here is put the new symbols in `sources` in their own crate then reference it from `testsys` in `tools`. This removes the copy of `tools` into the build container.

I should have done this initially, but I was enamored with the idea of the `buildsys` crate having to do with all things build related.

bcressey commented on sources/api/bootstrap-containers/build.rs:
```diff
@@ -1,32 +1,7 @@
-
-    let mut readme = File::create("README.md").unwrap();
-    readme.write_all(content.as_bytes()).unwrap();
+    generate_readme(ReadmeSource::Main).unwrap()
```
I don't mind this but it might be just as intuitive to have functions like:
```
generate_readme_main()
generate_readme_lib()
```

or
```
generate_readme::from_main()
generate_readme::from_lib()
```

(if it were a separate crate and we didn't want to have the import)

bcressey commented on sources/api/certdog/build.rs:
```diff
@@ -1,32 +1,7 @@
 // Automatically generate README.md from rustdoc.
```
We could potentially clean up these comments as well.

bcressey commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+/// will need to be rebuilt if this changes. `Makefile.toml` emits the variant string in the
+/// `BUILDSYS_VARIANT` environment variable. This is then passed to crate builds by the `Dockerfile`
+/// as `VARIANT`.
+pub const VARIANT_ENV: &str = "VARIANT";
```
This could also be a separate crate (`bottlerocket-variant`) and we could import it in `builder.rs` to ensure we set the expected `VARIANT_ENV` value.

For that matter it might be useful to export all of these other fields into the environment, so that spec files don't have to mimic the parsing (as `os.spec` does).

bcressey commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+
+/// The default `variant_type`. If the fourth position of a variant string tuple does not exist,
+/// then the variant is considered to be `"general_purpose"`.
+pub const DEFAULT_VARIANT_TYPE: &str = "general_purpose";
```
Can we omit these default values and just leverage the fact that the fields are `Option<String>`s and leave them set to `None`?

Otherwise we could default to more innocuous values, like `0` for the version and `"none"` for the type.

("general purpose" is a bit of an EC2-ism.)

bcressey commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+    variant: String,
+    platform: String,
+    runtime: String,
+    family: String,
```
I don't love `family` but I don't have a better suggestion. I think this ends up being a pretty important concept for us elsewhere, e.g. #1261.

I'm fine with this for now but we might want to revisit this name (and "variant" for that matter) before we use it more widely.

bcressey commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+        self.version.as_deref().unwrap_or(DEFAULT_VARIANT_VERSION)
+    }
+
+    /// The variant's type. This is the optional fourth value in the variant string tuple. For
+    /// example for `aws-k8s-1.21-nvidia` the `variant_type` is `nvidia`. If the `variant_type` does
+    /// not exist, [`DEFAULT_VARIANT_TYPE`] is returned.
+    pub fn variant_type(&self) -> &str {
+        self.variant_type.as_deref().unwrap_or(DEFAULT_VARIANT_TYPE)
+    }
```
Maybe bring back `flavor` for this concept? Naming is hard, but I'd like to get away from terminology with a counterpart in genetics.

bcressey reviewed (commented):
Thanks! Overall this looks like a nice win in terms of eliminating a lot copy/paste code and bringing some consistency to how we interact with variants programmatically.

arnaldo2792 commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+/// Provides a structure for representing a Bottlerocket variant. Provides Functionality useful in
+/// build scripts and other tooling that are variant-aware.
```
nit: should the two phrases be joined with `and`?

arnaldo2792 commented on tools/buildsys/src/public/variant.rs:
```diff
@@ -0,0 +1,412 @@
+        );
+        let variant_family = format!("{}-{}", platform, runtime);
+        let variant_version = parts.next().map(|s| s.to_string());
+        if let Some(value) = variant_version.as_ref() {
+            ensure!(
+                !value.is_empty(),
+                error::VariantPartEmptySnafu {
+                    part_name: "variant_version",
+                    variant: variant.clone()
+                }
+            );
+        }
+        let variant_type = parts.next().map(|s| s.to_string());
+        if let Some(value) = variant_type.as_ref() {
+            ensure!(
+                !value.is_empty(),
+                error::VariantPartEmptySnafu {
+                    part_name: "variant_type",
+                    variant: variant.clone()
+                }
+            );
+        }
```
Could you save yourself these two checks by adding a regex with the valid format of the variant? If `aws-k8s-` is given, the regex will catch the problem and the `is_empty()` check can be avoided.

webern replied:
I would prefer not to use regex just to check if the tuple position is empty. It looks like more code duplication than it actually is because 90% of it is error instantiation.

webern committed:
generate readme

We had readme generating code pasted into most of our build.rs file.
This has now been moved to a public function in a library so that the
code is not duplicated.

webern committed:
standardize variant concepts

We had several crates that interpreted the VARIANT env variable in
different ways in build.rs. This standardizes the treatment of the
VARIANT variable by centralizing it in a library.

webern commented:
^ I believe I have addressed all of the comments.

bcressey commented on sources/bottlerocket-variant/src/lib.rs:
```diff
@@ -0,0 +1,442 @@
+/// - metal-dev
+/// - aws-k8s-1.21-nvidia
+///
+/// The `platform` and `runtime` values are required. `variant_version` and `variant_flavor` values
+/// are optional and will default to `"undefined"` and `"general_purpose"` respectively.
+///
```
```suggestion
/// The `platform` and `runtime` values are required. `variant_version` and `variant_flavor` values
/// are optional and will default to `"0"` and `"none"` respectively.
///
```

webern commented:
I reran testing for aws-k8s, vmware-k8s and aws-ecs
